### DISCUSSION TITLE
Using MikroORM as library generates 2 hydrators?

### DISCUSSION THREAD

**Discussion by Tlepel** [mikro-orm/mikro-orm#4304] [Q&A]
We were looking to upgrade MikroORM versions. At the moment, we're using 5.6.7, but we need the fix for #4245 before we can finish a new feature. However, after the update our automated tests failed repeatedly.

I delved into the issue at hand, and it seemed like a simple `findOne` fetched an entity and created a changed entity at the same time.

```ts
@Entity()
class A {
  @PrimaryKey()
  id!: number;

  @OneToOne(() => B, b => b.a, {
    orphanRemoval: true,
    eager: false,
    wrappedReference: true,
  })
  b!: IdentifiedReference<B>;

  @Property()
  value!: string;
}

@Entity()
class B {
  @PrimaryKey()
  id!: number;

  @Property()
  value!: string;

  @OneToOne({
    entity: () => A,
    inversedBy: 'b',
    wrappedReference: true,
    serializer: a => a.id,
    onDelete: 'cascade',
  })
  a!: IdentifiedReference<A>;

  @OneToMany(
    () => C,
    c => c.parent,
    {
      eager: true,
      orphanRemoval: true,
    },
  )
  children = new Collection<C>(this);
}

@Entity()
class C {
  @PrimaryKey()
  id!: number;

  @Property()
  value!: string;

  @ManyToOne(() => B, {
    serializer: b => b.id,
    wrappedReference: true,
    onDelete: 'cascade',
  })
  parent!: IdentifiedReference<B>;
}
```

With this configuration, from the repository for `A`, we do `repository.findOne(1)`, which returns an `A` entity. However, at that moment the changeset contains an entry for `B`, with payload `{a: 1}`. This causes problems after calling `em.flush()`, since `a.b.unwarp().children` is undefined, and we cannot load these children.

I did some digging, and it turns out that the `ObjectHydrator` actually has 2 instances; the `running` state is changed on the 2nd instance during hydration, while the hydrator in the `EntityHelper` is the other instance. Therefore, the property setter for the `IdentifiedReference<A>` finds that the hydrator isn't running, and therefore the `__originalEntityData` won't be populated with the initial state of the entity. Therefore the changeset will include it as a change, since the diff includes the reference.

This behaviour changed since 5.6.10; when using 5.6.9 the tests pass and this behaviour isn't present. I couldn't find any relevant changes in the diffs, however, but that's probably because I'm not super familiar with the implementation.

I couldn't reproduce this in a test, however, since it probably has something to do with the configuration. 

I'm looking for some help in brainstorming where the config could be different. Our config doesn't include the `hydrator` property, so the default `ObjectHydrator` is used by MikroORM.

If you can help, please let me know :)

**Comment by B4nan** [comment]
>  we're using 5.6.7

Have you tried latest version?

There should be only one hydrator instance, `EntityHelper` does not create its own - it uses `em.getHydrator()` which uses the cached instance from configuration object.

**Reply by Tlepel** [reply]
Yes, however this problem only appeared came up when we upgraded to the lates version (5.7.x). I backtracked and saw that it was introduced in 5.6.10. Because of this we haven't upgraded.

I tried to reproduce the issue, however in the mikro-orm repo it works as expected.

**Reply by B4nan** [reply]
Then try to reproduce it outside, in a dedicated repository. 

If you have two hydrators, it could be as well an issue with having multiple core packages installed in your project. That often happens in wrongly configured monorepos, maybe that's the problem?

**Reply by Tlepel** [reply]
Hmm, I thought about this and tried to reproduce it.. The behaviour changed in commit 8bb0268a6d67143b0aa4dd0a5c6a6fb1bd0f8374, we backtracked that using yarn link to build specific commits of MikroORM.

I actually managed to reproduce it in this repo, so I'll link to this discussion in the failing test.
